src/backend/replication/walproposer.c

```diff
@@ -443,14 +546,50 @@ WalProposerMain(Datum main_arg)
 		ReplicationSlotRelease();
 	}
 
-	/* Initiate connections to all walkeeper nodes */
-	for (int i = 0; i < n_walkeepers; i++)
-	{
-		ResetConnection(i);
-	}
+	WalProposerStart();
+}
 
-	while (true)
-		WalProposerPoll();
+void
+WalProposerSync(int argc, char *argv[])
+{
+	syncSafekeepers = true;
+
+	InitStandaloneProcess(argv[0]);
+
+	SetProcessingMode(InitProcessing);
+
+	/*
+	 * Set default values for command-line options.
+	 */
+	InitializeGUCOptions();
+
+	/* Acquire configuration parameters */
+	if (!SelectConfigFiles(NULL, progname))
+		exit(1);
+
+	process_shared_preload_libraries_in_progress = true;
```

Comment:
Is it here to indicate that WalProposerInit() loads libraries using load_file()?

If so, then maybe move this into WalProposerInit() itself? Otherwise, e.g. in the WalProposerMain() it is called without this `process_shared_preload_libraries_in_progress = true / false` wrapper

CC @arssher @knizhnik

Reply:
No, `process_shared_preload_libraries_in_progress` flag is set by Potgresitslef while loading shared libraries.
It is set here before calling WalProposerInit() because otherwise attempt to load dynamic libraries will cause error when WalProposerInit()  is called not as part of standard bgworker startup,  but directly from PostgreMain when --sync-walkeepers option is specified. In other words by setting this flags we simulate  preloading of shared libraries.

Reply:
I've added a comment on that.